File: src/main/java/net/guizhanss/fastmachines/implementation/listeners/HopperListener.java
package net.guizhanss.fastmachines.implementation.listeners;

import org.bukkit.Location;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryMoveItemEvent;
import org.bukkit.event.inventory.InventoryPickupItemEvent;
import org.bukkit.event.inventory.InventoryType;

import me.mrCookieSlime.Slimefun.api.BlockStorage;

import net.guizhanss.fastmachines.FastMachines;
import net.guizhanss.fastmachines.core.items.attributes.NotAHopper;
import net.guizhanss.fastmachines.utils.CompatUtils;

public class HopperListener implements Listener {

    public HopperListener(FastMachines plugin) {
        plugin.getServer().getPluginManager().registerEvents(this, plugin);
    }

    @EventHandler
    public void onPickupItem(InventoryPickupItemEvent e) {
        if (e.getInventory().getType() != InventoryType.HOPPER) {
            return;
        }

        // Inventory#getLocation() does not exist on 1.8.8 - resolve reflectively (null-safe there).
        Location loc = CompatUtils.inventoryLocation(e.getInventory());
        if (loc == null) {
            return;
        }

        if (BlockStorage.check(loc) instanceof NotAHopper) {
            e.setCancelled(true);
        }
    }

    @EventHandler
    public void onMoveItemIntoHopper(InventoryMoveItemEvent e) {
        if (e.getDestination().getType() != InventoryType.HOPPER) {
            return;
        }

        Location loc = CompatUtils.inventoryLocation(e.getDestination());
        if (loc == null) {
            return;
        }

        if (BlockStorage.check(loc) instanceof NotAHopper) {
            e.setCancelled(true);
        }
    }
}

File: src/main/kotlin/net/guizhanss/fastmachines/implementation/listeners/HopperListener.kt
package net.guizhanss.fastmachines.implementation.listeners

import me.mrCookieSlime.Slimefun.api.BlockStorage
import net.guizhanss.fastmachines.FastMachines
import net.guizhanss.fastmachines.core.items.attributes.NotAHopper
import net.guizhanss.fastmachines.utils.CompatUtils
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryMoveItemEvent
import org.bukkit.event.inventory.InventoryPickupItemEvent
import org.bukkit.event.inventory.InventoryType

class HopperListener(plugin: FastMachines) : Listener {

    init {
        plugin.server.pluginManager.registerEvents(this, plugin)
    }

    @EventHandler
    fun onPickupItem(e: InventoryPickupItemEvent) {
        if (e.inventory.type != InventoryType.HOPPER) return

        // Inventory#getLocation() does not exist on 1.8.8 - resolve reflectively (null-safe there).
        val loc = CompatUtils.inventoryLocation(e.inventory) ?: return

        val sfItem = BlockStorage.check(loc)
        if (sfItem is NotAHopper) {
            e.isCancelled = true
        }
    }

    @EventHandler
    fun onMoveItemIntoHopper(e: InventoryMoveItemEvent) {
        if (e.destination.type != InventoryType.HOPPER) return

        val loc = CompatUtils.inventoryLocation(e.destination) ?: return

        if (BlockStorage.check(loc) is NotAHopper) {
            e.isCancelled = true
        }
    }
}
